packages/agreement-process/src/services/validators.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  Attribute,
+  Attributes,
+  Descriptor,
+  DescriptorState,
+  EService,
+  PersistentAgreement,
+  PersistentAgreementState,
+  Tenant,
+  TenantAttribute,
+  WithMetadata,
+  agreementAlreadyExists,
+  agreementNotFound,
+  descriptorNotInExpectedState,
+  descriptorState,
+  missingCertifiedAttributesError,
+  notLatestEServiceDescriptor,
+  operationNotAllowed,
+  persistentAgreementState,
+  tenantAttributeType,
+} from "pagopa-interop-models";
+import { ApiAgreementPayload } from "../model/types.js";
+import { readModelService } from "./readModelService.js";
+
+const validateLatestDescriptor = (
+  eService: EService,
+  descriptorId: string,
+  allowedStates: DescriptorState[]
+): Descriptor => {
+  const recentActiveDescriptor = eService.descriptors
+    .filter((d) => d.state !== descriptorState.draft)
+    .sort((a, b) => Number(a.version) - Number(b.version))
+    .find((d) => d.id === descriptorId);
+
+  if (!recentActiveDescriptor) {
+    throw notLatestEServiceDescriptor(descriptorId);
+  }
+  validateDescriptorState(
+    eService.id,
+    descriptorId,
+    recentActiveDescriptor.state,
+    allowedStates
+  );
+
+  return recentActiveDescriptor;
+};
+
+const attributesSatisfied = (
+  requested: Attribute[][],
+  assigned: string[]
+): boolean =>
+  requested.every((attributeList) => {
+    const attributes = attributeList.map((a) => a.id);
+    return attributes.filter((a) => assigned.includes(a)).length > 0;
+  });
+
+const validateDescriptorState = (
+  eserviceId: string,
+  descriptorId: string,
+  descriptorState: DescriptorState,
+  allowedStates: DescriptorState[]
+): void => {
+  if (!allowedStates.includes(descriptorState)) {
+    throw descriptorNotInExpectedState(eserviceId, descriptorId, allowedStates);
+  }
+};
+
+const certifiedAttributesSatisfied = (
+  eServiceAttributes: Attributes,
+  consumerAttributes: TenantAttribute[]
+): boolean =>
+  attributesSatisfied(
+    eServiceAttributes.certified,
+    consumerAttributes
+      .filter((a) => a.type === "certified" && !a.revocationTimestamp)
```

Comment:
I guess the (duplicated) filter on `type` can be avoided if `consumerAttributes` type is changed to `CertifiedTenantAttribute` (defined as a separated type [here](https://github.com/pagopa/interop-be-monorepo/pull/96/files#diff-1536a0b824255a4f5f79848ede99764c14f84f3c4ec5c8030f00f3363abb1f60R55))

Reply:
In this case, merging the two filtering in `certifiedAttributesDescriptorSatisfied` could be the best solution

Reply:
it works for me 👍 , I would still ask you to define the types for the different attributes though.
It makes much clearer the correlation with the business requirement